### PR TITLE
SemVer 2.0.0 version ranges imply that the package itself is SemVer 2.0.0

### PR DESCRIPTION
**Pull request by joelverhagen**
Addresses https://github.com/NuGet/NuGetGallery/issues/3657.

This is adhering to the latest (internal) SemVer 2.0.0 protocol specification.

### CODE REVIEW

File: src/NuGet.Server.Core/Infrastructure/ServerPackage.cs [hunk 1]
```diff
@@ -249,5 +253,41 @@ private static List<PackageDependencySet> ParseDependencySet(string value)
 
             return Tuple.Create(id, versionSpec, targetFramework);
         }
+
+        private static bool IsPackageSemVer2(IPackage package)
```
Comment (xavierdecoster) [inline src/NuGet.Server.Core/Infrastructure/ServerPackage.cs]: This is exactly the method that we may need to re-use in gallery and search service/jobs. Is the intention for us to all consume NuGet.Server.Core (as a NuGet package)?

Comment (joelverhagen) [inline src/NuGet.Server.Core/Infrastructure/ServerPackage.cs]: I think NuGet.Server should be separate. It depends on NuGet.Core (`IPackage`). Also, there are currently no dependencies between gallery and NuGet.Server repositories.

I think the sharing should be done amongst NuGet.org components, but not with NuGet.Server.

Comment (xavierdecoster) [inline src/NuGet.Server.Core/Infrastructure/ServerPackage.cs]: Exactly the reason for my question :) Perfect.

**Comment by joelverhagen** [comment]
Nope, we chose to limit the evaluation to just a single package's nuspec. It's​ a very hard problem to evaluate ever possible restore graph a package and it's transitive dependencies can be a part to check for any SemVer 2.0.0 version strings.

The failure experience on the client should either be a "dependency not found" error (if there is no SemVer 1.0.0 package satisfying the dependency range) or "exact match not found" warning (if the minimum bound is SemVer 2.0.0 and therefore ignored).

**Comment by maartenba** [comment]
But in that case the check should not be done on dependencies I guess? If the client errors out anyway, why bother with hidden complexity :-)

**Comment by joelverhagen** [comment]
> But in that case the check should not be done on dependencies I guess? If the client errors out anyway, why bother with hidden complexity :-)

The old clients will blow up with a parsing error if the version range contains SemVer 2.0.0. This is a worse experience than "package not found" on dependencies, IMO.

**Comment by maartenba** [comment]
That is true, got it.

**Pull request merged**